bin/PrintHighestContainingSubstitution/Main.cpp

```diff
@@ -0,0 +1,171 @@
+#include <fcntl.h>
+#include <pasta/AST/AST.h>
+#include <pasta/AST/Macro.h>
+#include <pasta/Compile/Command.h>
+#include <pasta/Compile/Compiler.h>
+#include <pasta/Compile/Job.h>
+#include <pasta/Util/ArgumentVector.h>
+#include <pasta/Util/File.h>
+#include <pasta/Util/FileSystem.h>
+#include <pasta/Util/Init.h>
+#include <unistd.h>
+
+#include <csignal>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+
+void PrintTokensTo(std::iostream &s, const pasta::TokenRange &tokens) {
+  int i = 0;
+  for (auto &&token : tokens) {
+    if (token.Role() == pasta::TokenRole::kFinalMacroExpansionToken ||
+        token.Role() == pasta::TokenRole::kFileToken) {
+      if (i > 0) {
+        s << ' ';
+      }
+      s << token.Data();
+      i++;
+    }
+  }
+}
+
+class StmtPrintHighestContainingSubstitutionFinder final : public pasta::StmtVisitor {
+public:
+  std::ostream &os;
+
+  virtual ~StmtPrintHighestContainingSubstitutionFinder(void) = default;
+
+  explicit StmtPrintHighestContainingSubstitutionFinder(std::ostream &os_) : os(os_) {}
+
+  void VisitStmt(const pasta::Stmt &stmt) {
+    if (stmt.Kind() != pasta::StmtKind::kImplicitCastExpr
```

Comment:
Why these conditions? Is this because they wrap other nodes, but have identical bounds as the other nodes? If so, add a comment.